### PR TITLE
text_classification (sst-2) fix

### PR DESCRIPTION
**Pull request by ekmb**
1. fixed a bug in text_classification script for sst-2 dataset
2. fixed shuffling in the text_dataset/layer
3. cleaned up text_classification_with_bert.ipynb
4. moved TextDatasetDesc out of textDataset.py and moved both to a folder (was broken during merge)

update:
5. change BertTokenizer to NemoBertTokenizer
6. added roberta and albert support to text_classification
7. added function to create sentence piece tokenizer with special tokens
8. test for SPT updated

Signed-off-by: Evelina Bakhturina <ebakhturina@nvidia.com>

### CODE REVIEW

File: nemo/collections/nlp/data/datasets/text_classification/text_classification_dataset.py [hunk 1]
```diff
@@ -62,14 +48,13 @@ class BertTextClassificationDataset(Dataset):
     """
 
     def __init__(
-        self, input_file, max_seq_length, tokenizer, num_samples=-1, shuffle=True, use_cache=False,
+        self, input_file, max_seq_length, tokenizer, num_samples=-1, use_cache=False,
```
[the remark this answers is not in the document]
Comment (ekmb) [reply nemo/collections/nlp/data/datasets/text_classification/text_classification_dataset.py]: done

File: examples/nlp/text_classification/text_classification_with_bert.py [hunk 2]
```diff
@@ -113,7 +111,7 @@ def create_pipeline(num_samples=-1, batch_size=32, num_gpus=1, local_rank=0, mod
     data_file = f'{data_desc.data_dir}/{mode}.tsv'
     shuffle = args.shuffle_data if mode == 'train' else False
 
-    data_layer = nemo.collections.nlp.nm.data_layers.text_classification_datalayer.BertTextClassificationDataLayer(
+    data_layer = nemo_nlp.nm.data_layers.text_classification_datalayer.BertTextClassificationDataLayer(
```
[the remark this answers is not in the document]
Comment (ekmb) [reply examples/nlp/text_classification/text_classification_with_bert.py]: done

File: examples/nlp/text_classification/text_classification_with_bert.py [hunk 3]
```diff
@@ -78,26 +78,24 @@
 
 """ Load the pretrained BERT parameters
 See the list of pretrained models, call:
-nemo_nlp.huggingface.BERT.list_pretrained_models()
+nemo.collections.nlp.nm.trainables.huggingface.BERT.list_pretrained_models()
```
[the remark this answers is not in the document]
Comment (ekmb) [reply examples/nlp/text_classification/text_classification_with_bert.py]: done

File: examples/nlp/text_classification/text_classification_with_bert.py [hunk 4]
```diff
@@ -78,26 +78,24 @@
...
 
 hidden_size = pretrained_bert_model.hidden_size
 tokenizer = BertTokenizer.from_pretrained(args.pretrained_bert_model)
 
-data_desc = TextClassificationDataDesc(args.dataset_name, args.data_dir, args.do_lower_case)
+data_desc = TextClassificationDataDesc(args.dataset_name, args.data_dir, args.do_lower_case, args.eval_file_prefix)
 
 # Create sentence classification loss on top
-classifier = nemo.collections.nlp.nm.trainables.common.sequence_classification_nm.SequenceClassifier(
+classifier = nemo_nlp.nm.trainables.sequence_classification_nm.SequenceClassifier(
```
[the remark this answers is not in the document]
Comment (ekmb) [reply examples/nlp/text_classification/text_classification_with_bert.py]: done

File: examples/nlp/text_classification/text_classification_with_bert.py [hunk 5]
```diff
@@ -78,26 +78,24 @@
...
 
 if args.bert_checkpoint and args.bert_config:
-    pretrained_bert_model = nemo.collections.nlp.nm.trainables.common.huggingface.BERT(
-        config_filename=args.bert_config
-    )
+    pretrained_bert_model = nemo.collections.nlp.nm.trainables.huggingface.BERT(config_filename=args.bert_config)
     pretrained_bert_model.restore_from(args.bert_checkpoint)
 else:
-    pretrained_bert_model = nemo.collections.nlp.nm.trainables.common.huggingface.BERT(
+    pretrained_bert_model = nemo.collections.nlp.nm.trainables.huggingface.BERT(
```
[the remark this answers is not in the document]
Comment (ekmb) [reply examples/nlp/text_classification/text_classification_with_bert.py]: done

File: examples/nlp/glue_benchmark/glue_benchmark_with_bert.py [hunk 6]
```diff
@@ -248,6 +246,16 @@
     pooler = SequenceClassifier(hidden_size=hidden_size, num_classes=num_labels, log_softmax=False)
     glue_loss = CrossEntropyLossNM()
 
+if args.bert_checkpoint is not None:
+    model.restore_from(args.bert_checkpoint)
+    logging.info(f"model restored from {args.bert_checkpoint}")
+
+    if args.task_head_checkpoint is not None:
```
Comment (VahidooX) [inline examples/nlp/glue_benchmark/glue_benchmark_with_bert.py]: Isn't task_head_checkpoint param independent of bert_checkpoint? Shouldn't this be out of the block?

File: examples/nlp/text_classification/text_classification_with_bert.py [hunk 7]
```diff
@@ -76,31 +78,45 @@
...
+hidden_size = model.hidden_size
+
+tokenizer_cls = nemo_nlp.data.NemoBertTokenizer
+tokenizer_special_tokens = nemo_nlp.utils.MODEL_SPECIAL_TOKENS[model_type]
+tokenizer = tokenizer_cls(
+    pretrained_model=args.pretrained_model_name, special_tokens=tokenizer_special_tokens, bert_derivate=model_type,
+)
 
-data_desc = TextClassificationDataDesc(args.dataset_name, args.data_dir, args.do_lower_case)
+data_desc = TextClassificationDataDesc(args.dataset_name, args.data_dir, args.do_lower_case, args.eval_file_prefix)
```
Comment (VahidooX) [inline examples/nlp/text_classification/text_classification_with_bert.py]: How about passing the list of prefixes to peocess? like ['train', args.eval_file_prefix]
So that it is more flexible.

File: examples/nlp/text_classification/text_classification_with_bert.py [hunk 8]
```diff
@@ -76,31 +78,45 @@
...
+    nemo_nlp.nm.trainables.huggingface.BERT.list_pretrained_models()
+    nemo_nlp.nm.trainables.huggingface.Albert.list_pretrained_models()
+    nemo_nlp.nm.trainables.huggingface.Roberta.list_pretrained_models()
+    """
+    model = model_cls(pretrained_model_name=args.pretrained_model_name)
+
+hidden_size = model.hidden_size
+
+tokenizer_cls = nemo_nlp.data.NemoBertTokenizer
+tokenizer_special_tokens = nemo_nlp.utils.MODEL_SPECIAL_TOKENS[model_type]
```
Comment (VahidooX) [inline examples/nlp/text_classification/text_classification_with_bert.py]: How about moving MODEL_SPECIAL_TOKENS into the tokenizer and pass the model_type to it? @yzhang123

File: examples/nlp/text_classification/text_classification_with_bert.py [hunk 9]
```diff
@@ -113,14 +129,15 @@ def create_pipeline(num_samples=-1, batch_size=32, num_gpus=1, local_rank=0, mod
...
+    data_layer = nemo_nlp.nm.data_layers.BertTextClassificationDataLayer(
         input_file=data_file,
         tokenizer=tokenizer,
         max_seq_length=args.max_seq_length,
         num_samples=num_samples,
         shuffle=shuffle,
         batch_size=batch_size,
-        use_cache=args.use_cache,
+        use_cache=not args.no_data_cache,
+        model_name=args.pretrained_model_name,
```
Comment (VahidooX) [inline examples/nlp/text_classification/text_classification_with_bert.py]: Why does data_layer need model_name?

File: nemo/collections/nlp/data/datasets/text_classification/text_classification_data_descriptor.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+            self.data_dir = process_nlu(data_dir, do_lower_case, dataset_name=dataset_name)
+            self.eval_file = self.data_dir + '/test.tsv'
+        elif dataset_name.startswith('jarvis'):
+            self.data_dir = process_jarvis_datasets(
+                data_dir, do_lower_case, dataset_name, modes=['train', 'test', 'eval'], ignore_prev_intent=False
+            )
+
+            intents = get_intent_labels(f'{self.data_dir}/dict.intents.csv')
+            self.num_labels = len(intents)
+        else:
```
Comment (VahidooX) [inline nemo/collections/nlp/data/datasets/text_classification/text_classification_data_descriptor.py]: What happens when user wants to try their own dataset? Do we have a format that they can use?

Comment (ekmb) [reply nemo/collections/nlp/data/datasets/text_classification/text_classification_data_descriptor.py]: added default_format

File: examples/nlp/text_classification/text_classification_with_bert.py (hunk 9 above)
Comment (ekmb) [reply examples/nlp/text_classification/text_classification_with_bert.py]: different model use different tokenizers, so we need to add model name to the cache file name to distinguish preprocess datatsets

File: examples/nlp/text_classification/text_classification_with_bert.py (hunk 8 above)
Comment (ekmb) [reply examples/nlp/text_classification/text_classification_with_bert.py]: done

File: examples/nlp/text_classification/text_classification_with_bert.py (hunk 7 above)
Comment (ekmb) [reply examples/nlp/text_classification/text_classification_with_bert.py]: done

File: nemo/collections/nlp/nm/data_layers/text_classification_datalayer.py [hunk 11]
```diff
@@ -76,7 +78,6 @@ def __init__(
             'tokenizer': tokenizer,
             'max_seq_length': max_seq_length,
             'num_samples': num_samples,
-            'shuffle': shuffle,
```
[the remark this answers is not in the document]
Comment (ekmb) [reply nemo/collections/nlp/nm/data_layers/text_classification_datalayer.py]: this is required for the first shuffle

File: examples/nlp/question_answering/question_answering_squad.py [hunk 12]
```diff
@@ -329,38 +325,18 @@ def create_pipeline(
...
-    else:
-        tokenizer_cls = nemo_nlp.data.NemoBertTokenizer
-        tokenizer_special_tokens = nemo_nlp.utils.MODEL_SPECIAL_TOKENS[model_type]
-        tokenizer = tokenizer_cls(
-            pretrained_model=args.pretrained_model_name,
-            special_tokens=tokenizer_special_tokens,
-            bert_derivate=model_type,
+    if args.tokenizer == 'sentencepiece':
+        tokenizer = nemo.collections.nlp.data.tokenizers.huggingface_utils.get_sentence_piece_tokenizer(
+            tokenizer_model=args.tokenizer_model, pretrained_model_name=args.pretrained_model_name
```
Comment (yzhang123) [inline examples/nlp/question_answering/question_answering_squad.py]: could you set tokenizer_model in argparse to None for consistency with other models?

File: examples/nlp/text_classification/text_classification_with_bert.py (hunk 9 above)
Comment (yzhang123) [reply examples/nlp/text_classification/text_classification_with_bert.py]: could you change all datalayers to have this feature for consistency?

File: nemo/collections/nlp/nm/trainables/common/huggingface/huggingface_utils.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+            return MODELS[model_type]['class'](config_filename=bert_config)
+        else:
+            return MODELS[model_type]['class'](pretrained_model_name=pretrained_model_name)
+    else:
+        raise ValueError(f'Choose pretrained model from the following list: {get_huggingface_models_list()}.')
+
+
+MODELS = {
+    'bert': {'default': 'bert-base-uncased', 'class': BERT},
+    'roberta': {'defualt': 'roberta-base', 'class': Roberta},
```
Comment (yzhang123) [inline nemo/collections/nlp/nm/trainables/common/huggingface/huggingface_utils.py]: typo

File: nemo/collections/nlp/nm/trainables/common/huggingface/huggingface_utils.py [hunk 14]
```diff
@@ -0,0 +1,55 @@
+# =============================================================================
```
Comment (yzhang123) [inline nemo/collections/nlp/nm/trainables/common/huggingface/huggingface_utils.py]: why did you move this to nm.trainables, and not leave it in utils?

File: examples/nlp/question_answering/question_answering_squad.py [hunk 15]
```diff
@@ -329,38 +325,18 @@ def create_pipeline(
...
-        tokenizer.add_special_tokens(special_tokens)
-    else:
-        tokenizer_cls = nemo_nlp.data.NemoBertTokenizer
-        tokenizer_special_tokens = nemo_nlp.utils.MODEL_SPECIAL_TOKENS[model_type]
-        tokenizer = tokenizer_cls(
-            pretrained_model=args.pretrained_model_name,
-            special_tokens=tokenizer_special_tokens,
-            bert_derivate=model_type,
+    if args.tokenizer == 'sentencepiece':
+        tokenizer = nemo.collections.nlp.data.tokenizers.huggingface_utils.get_sentence_piece_tokenizer(
```
Comment (yzhang123) [inline examples/nlp/question_answering/question_answering_squad.py]: can we change it to nemo_nlp.data ...?

File: examples/nlp/glue_benchmark/glue_benchmark_with_bert.py [hunk 16]
```diff
@@ -16,7 +16,7 @@
 # limitations under the License.
 # =============================================================================
 
-"""
+'''
```
[the remark this answers is not in the document]
Comment (ekmb) [reply examples/nlp/glue_benchmark/glue_benchmark_with_bert.py]: done

File: examples/nlp/question_answering/question_answering_squad.py (hunk 12 above)
Comment (ekmb) [reply examples/nlp/question_answering/question_answering_squad.py]: done

File: nemo/collections/nlp/nm/trainables/common/huggingface/huggingface_utils.py (hunk 13 above)
Comment (ekmb) [reply nemo/collections/nlp/nm/trainables/common/huggingface/huggingface_utils.py]: fixed

Comment (ekmb) [reply nemo/collections/nlp/nm/trainables/common/huggingface/huggingface_utils.py]: regarding "could you change all datalayers to have this feature for consistency?" - let's make a separate PR for this + we need to decide on pickle vs hdf5

File: nemo/collections/nlp/nm/trainables/common/huggingface/huggingface_utils.py (hunk 14 above)
Comment (ekmb) [reply nemo/collections/nlp/nm/trainables/common/huggingface/huggingface_utils.py]: had import loop issues and this file is needed only for huggingface trainables

File: examples/nlp/question_answering/question_answering_squad.py (hunk 15 above)
Comment (ekmb) [reply examples/nlp/question_answering/question_answering_squad.py]: changed to data.tokenizers()
